### DISCUSSION TITLE
`export const defer = i`

### DISCUSSION THREAD

**Discussion by boehs** [withastro/roadmap#127] [Proposal]
With @tony-sull 's wonderful non HTML support, it's time to revisit deferring compilation

```js
export const defer = 2
```

All the pages are `defer = 0` by default, meaning they get built first. Then `defer = 1`, `defer = 2`, etc. If astro sees a page with defer in the frontmatter it adds the page to a backlog and moves onto the next page

```js
// the backlog object
{
1: ["path/to/page/"],
2: ["these/pages/have", "a/defer/value/of/2"]
}
```

## Examples

This is useful for many things, examples:

- Waiting for a HTML page to be built, then opening it from dist using puppeteer to generate a SVG OG image
- pages relient on side effects of building other pages (like a sitemap)
  - Or collecting data in a layout (date, post title, content), and exporting that data in a search friendly way, for adding a search component to the site.

### Concrete minimal example

Some people like to get their hands dirty, or maybe I'm bad at describing things! 

https://stackblitz.com/edit/astro-deferred-compilation-usecase?file=src%2Fpages%2Findex.astro

## Issues

Because we don't want to run the intensive build task for a page that we don't need to yet, rules will apply to make it easy to parse before an AST is generated (using regex, perhaps)

The file would need to explicitly have in the frontmatter the following string

```js
export const defer = 
```
There are rules
- `defer` must be constant
- `defer` must be initialized with an integer and only an integer, no functions or variables allowed
- `defer` must be defined in frontmatter
- `defer` will only be read if it is in a file that is a child of the `pages` directory (no components or layouts)

This helps avoid executing javascript to find the value of defer, and saves the step of making an AST before it's needed.

**Comment by FredKSchott** [comment]
## Notes

- Nate: tabindex problem, using a single integer can be a limited API for what you're trying to describe
- Nate: In SSR, you don't know all pages at build time, because you are dynamically responding to requests in the server.
- Nate: for the sitemap problem, can we expose a manifest instead? Give you a better primitive JS object/data to build your own sitemap from. Prototype: `import {manifest} from "astro/manifest"`
